src/qibo/models/encodings.py

```diff
@@ -1600,135 +1632,436 @@ def _binary_encoder_hopf(
+        for word in cwres:
+
+            hw = hamming_weight(int(word))
+
+            inserted = False
+            for i in range(len(cw) - 1):
+                wi, wj = weights[i], weights[i + 1]
+                if wi <= hw <= wj:
+                    if (
+                        hamming_distance(int(word), int(cw[i])) <= 2
+                        and hamming_distance(int(word), int(cw[i + 1])) <= 2
+                    ):
+                        cw = backend.np.insert(cw, i + 1, word, axis=0)
+                        weights = backend.np.insert(weights, i + 1, hw)
+                        inserted = True
+                        break
+            if not inserted:
+                # append if no suitable interior gap is found
+                cw = backend.np.hstack((cw, word))
+                weights = backend.np.hstack((weights, hw))
+
+    return cw
+
+
+def _binary_codewords_ehrlich(dims: int, backend=None):
+    """
+    Yield fixed-width bitstrings representing integers 0..d-1, arranged so that
+    Hamming weights are strictly nondecreasing. Uses at most hamming_weight(d) calls
+    to `_ehrlich_codewords_up_to_k(k)`; when d is a power of two, exactly one call.
+    """
+    backend = _check_backend(backend)
+
+    # Check power-of-two case: b == 2^k
+    if (dims & (dims - 1)) == 0:
+        k = dims.bit_length() - 1  # since 2^k has bit_length k+1
+        # get_ehrlich_codewords(k) yields k-bit strings for 0..2^k-1 == 0..d-1
+        yield from _ehrlich_codewords_up_to_k(k, False, backend=backend)
+        return
+
+    # General case
+    n = int(
+        max(1, backend.np.ceil(backend.np.log2(dims)))
+    )  # width so that 2^(n-1) < b < 2^n
+    # Bits of b in most-significant-bit → least-significant-bit order
+    bits = [(dims >> i) & 1 for i in range(n - 1, -1, -1)]
+    prefix = ""
+    flip = False
+    for i, bit in enumerate(bits):
+        k = n - 1 - i
+        if bit == 1:
+            # emit the full 0-prefixed block of width k, using Ehrlich order
+            # flip every other time to keep consecutive codes close
+            for suffix in _ehrlich_codewords_up_to_k(k, flip, backend=backend):
+                yield prefix + "0" + suffix
+            flip = not flip
+            prefix += "1"
+        else:
+            prefix += "0"
+    # no final singleton: we've generated exactly d bitstrings (0..d-1)
+
+
+def _ehrlich_codewords_up_to_k(up2k: int, reversed_list: bool = False, backend=None):
+    """
+    Yield all bitstrings with monotonically changing Hamming weight from 0..up2k (or the reverse),
+    such that consecutive strings have Hamming distance ≤ 2. Codewords for each weight are produced
+    by `_ehrlich_algorithm(initial_string, False)` (assumed to return an iterable of bitstrings).
+
+    Args:
+        up2k: Length of the bitstrings and the maximum Hamming weight.
+        reversed_list: If True, generate from weight up2k down to 0; otherwise 0 up to up2k.
```

Comment:
```suggestion
        up2k (int): Length of the bitstrings and the maximum Hamming weight.
        reversed_list (bool, optional): If ``True``, generate from weight ``up2k`` down to :math:`0`.
            Otherwise, from :math:`0` up to ``up2k``.
```